src/policy/ancestor_packages.h

```diff
@@ -0,0 +1,166 @@
+                    }
+                    return ancestor_subset.size() < rhs.ancestor_subset.size();
+                } else {
+                    return mining_sequence.value() < rhs.mining_sequence.value();
+                }
+            }
+        }
+    };
+    /** Map from each txid to PackageEntry */
+    std::map<uint256, PackageEntry> m_txid_to_entry;
+
+    /** Linearized transactions. Always topologically sorted (IsSorted()). If fee information is
+     * provided through LinearizeWithFees(), using mining_sequence scores. */
+    std::vector<std::reference_wrapper<PackageEntry>> m_txns;
+
+    /** Helper function for recursively constructing ancestor caches in ctor. */
+    void visit(const CTransactionRef&);
+public:
+    /** Constructs ancestor package, sorting the transactions topologically and constructing the
+     * txid_to_tx and ancestor_subsets maps. It is ok if the input txns is not sorted.
+     * Expects:
+     * - No duplicate transactions.
+     * - No conflicts between transactions.
+     */
+    AncestorPackage(const Package& txns_in);
+
+    bool IsAncestorPackage() const { return m_ancestor_package_shaped; }
+
+    /** Returns all of the transactions, linearized. */
+    Package Txns() const;
+
+    /** Returns all of the transactions, without the skipped and dangling ones, linearized. */
+    Package FilteredTxns() const;
+
+    /** Get the sorted, filtered ancestor subpackage for a tx. Includes the tx. Does not include
+     * ancestors that are MEMPOOL. If this tx or any ancestors are DANGLING, returns std::nullopt. */
+    std::optional<std::vector<CTransactionRef>> FilteredAncestorSet(const CTransactionRef& tx) const;
+
+    /** Get the total fee and vsize of the ancestor subpackage for a tx. Includes the tx. Does not
+     * include ancestors that are MEMPOOL. If this tx or any ancestors are DANGLING, returns
+     * std::nullopt. This result is always consistent with FilteredAncestorSet(). */
+    std::optional<std::pair<CAmount, int64_t>> FilteredAncestorFeeAndVsize(const CTransactionRef& tx) const;
+
+    /** Get the fee and vsize of a tx. Returns std::nullopt if this information is unknown or this
+     * tx is DANGLING. */
+    std::optional<std::pair<CAmount, int64_t>> GetFeeAndVsize(const CTransactionRef& tx) const;
+
+    /** Should be called if a tx or same txid transaction is found in mempool or submitted to it.
+     * From now on, skip this tx from any result in FilteredAncestorSet(). Does not affect Txns().
+     */
+    void Skip(const CTransactionRef& transaction);
+
+    /** Should be called if a tx will not be considered because it is missing inputs.
+     * Skip a transaction and all of its descendants. From now on, if this transaction is present
+     * in the ancestor set, FilteredAncestorSet() returns std::nullopt for that tx. Does not affect Txns(). */
+    void SkipWithDescendants(const CTransactionRef& transaction);
```

Comment:
Agree with making these names hint at intended usage instead of internal details, changing

Just realized that maybe `TX_CONFLICT` should be in the "skip" instead of "skip with descendants" category. Imagine parent+child package, parent already confirmed and child hasn't been received yet. `SkipWithDescendants` means the child gets skipped too... assuming we can see a UTXO from the parent in the coinsview, we should treat it similarly to how we treat same-txid-different-witness-already-in-mempool right?

Reply:
`TX_CONFLICT` happens in three cases:
1) wtxid in mempool already `txn-already-in-mempool`
2) txid in mempool already `txn-same-nonwitness-data-in-mempool`
3) already in chain `txn-already-known`

Indeed the third case could slip through with current logic. We should probably treat this similarly to the first two cases.

I'm still lobbying for `TX_PREMATURE_SPEND` to be moved to this block :)

Reply:
Ok makes sense. I will do this next week, need to add a different `MempoolAcceptResult::ResultType` I think, and will write some tests. Addressed the other comments though.

Reply:
Added, and moved `TX_PREMATURE_SPEND` into the dangle category. I decided not to make a new `ResultType` because the result itself isn't that different from the INVALID case. Also added a table to `MempoolAcceptResult` to illustrate what fields/properties each of the result types should have.